### PR TITLE
Minitest support

### PR DESCRIPTION
**Pull request by blaknite**
This PR takes the work done by @mariovisic in #86 and applies it on top of the current main branch. In the process I have done a some refactoring to make it compatible with the revised gem structure.

Minitest support can be enabled with:

```ruby
require "buildkite/collector"
Buildkite::Collector.configure(hook: :minitest)
```

closes #86

### CODE REVIEW

**Comment by swebb** [comment]
> Minitest support can be enabled with:
> 
> ```ruby
> require "buildkite/collector"
> Buildkite::Collector.configure(hook: :minitest)
> ```

Just want to mention that we probably want to include the token in that ^

**Comment by swebb** [comment]
For some reason the logger level is getting set to "DEBUG". Example output:

```
minitest❯ bundle exec ruby -Ilib:test tests_meme.rb
2022-05-31T08:13:13.605141881Z pid=761738 tid=1680 DEBUG: hello from main thread
2022-05-31T08:13:15.202427758Z pid=761738 tid=1680 DEBUG: starting socket connection process
2022-05-31T08:13:16.496417468Z pid=761738 tid=1700 DEBUG: listening in on socket
2022-05-31T08:13:16.502671513Z pid=761738 tid=1700 DEBUG: received welcome
2022-05-31T08:13:16.802389160Z pid=761738 tid=1700 DEBUG: received confirm_subscription
2022-05-31T08:13:16.802635531Z pid=761738 tid=1680 INFO: Connected to Buildkite Test Analytics!
2022-05-31T08:13:16.802701823Z pid=761738 tid=1680 DEBUG: connected
2022-05-31T08:13:16.803110251Z pid=761738 tid=1720 DEBUG: hello from write thread
Run options: --seed 57712

# Running:

F2022-05-31T08:13:16.811866141Z pid=761738 tid=1680 DEBUG: added b2436b19-96cc-437a-86d6-c48018b6c9e2 to send queue
.2022-05-31T08:13:16.813031568Z pid=761738 tid=1680 DEBUG: added 4fc478fe-ee62-411f-8ade-88b7477f6d7a to send queue
S2022-05-31T08:13:16.813964550Z pid=761738 tid=1680 DEBUG: added 172446b9-5314-4ec6-9615-82881ea39817 to send queue


Finished in 0.004946s, 606.5449 runs/s, 606.5449 assertions/s.

  1) Failure:
TestMeme#test_that_kitty_can_eat [tests_meme.rb:18]:
Expected: "OHAI!"
  Actual: nil

3 runs, 3 assertions, 1 failures, 0 errors, 1 skips

You have skipped tests. Run with --verbose for details.
2022-05-31T08:13:16.815154299Z pid=761738 tid=1680 DEBUG: closing socket connection
2022-05-31T08:13:16.815228396Z pid=761738 tid=1680 DEBUG: added EOT to send queue
2022-05-31T08:13:16.815777983Z pid=761738 tid=1680 DEBUG: Waiting for Buildkite Test Analytics to send results...
2022-05-31T08:13:16.815837125Z pid=761738 tid=1680 DEBUG: waiting for last confirm
2022-05-31T08:13:16.880146046Z pid=761738 tid=1700 DEBUG: received ping
2022-05-31T08:13:17.519117590Z pid=761738 tid=1700 DEBUG: received confirm for indentifiers: ["b2436b19-96cc-437a-86d6-c48018b6c9e2", "4fc478fe-ee62-411f-8ade-88b7477f6d7a", "172446b9-5314-4ec6-9615-82881ea39817"]
2022-05-31T08:13:17.519203983Z pid=761738 tid=1700 DEBUG: all identifiers have been confirmed
2022-05-31T08:13:17.519372926Z pid=761738 tid=1680 DEBUG: socket close
2022-05-31T08:13:17.519651345Z pid=761738 tid=1680 DEBUG: socket disconnect
2022-05-31T08:13:18.133844215Z pid=761738 tid=1700 ERROR: end of file reached
2022-05-31T08:13:18.134085745Z pid=761738 tid=1680 INFO: Buildkite Test Analytics completed
2022-05-31T08:13:18.134149445Z pid=761738 tid=1680 DEBUG: socket connection closed
```

I *think* the reason is [this line](https://github.com/buildkite/rspec-buildkite-analytics/blob/68b0a3969c7ab4f42a00bd649434ca82e29571b4/lib/buildkite/collector.rb#L68). I think `level` should be a keyword argument (i.e., `Buildkite::Collector::Logger.new($stderr, level)`. The relevant docs are [here](https://ruby-doc.org/stdlib-2.4.0/libdoc/logger/rdoc/Logger.html#class-Logger-label-How+to+close+a+logger).

The only problem I've got is I can't trigger the problem with rspec. Only with minitest and I'm not sure why.

File: spec/collector/minitest_plugin/reporter_spec.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+require "minitest"
+require "buildkite/collector/minitest_plugin/reporter"
+require "buildkite/collector/uploader"
+
+RSpec.describe Buildkite::Collector::MinitestPlugin::Reporter do
```
[the remark this answers is not in the document]
Comment (blaknite) [reply spec/collector/minitest_plugin/reporter_spec.rb:7]: Yeah, this is just testing that `reporter.record` doesn't raise an error.

File: spec/collector/minitest_plugin/reporter_spec.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  it "test reporter works with a passed minitest result" do
+    Buildkite::Collector.configure(
+      token: "fake",
+      url: "http://fake.buildkite.localhost/v1/uploads",
+      hook: :minitest
+    )
+    io = StringIO.new
+    reporter = Buildkite::Collector::MinitestPlugin::Reporter.new(io, {})
+    a_example = double("Test")
+    trace = double("Trace")
```
[the remark this answers is not in the document]
Comment (blaknite) [reply spec/collector/minitest_plugin/reporter_spec.rb]: You're not, but I was. I've updated the test.

File: lib/buildkite/collector/minitest_plugin/trace.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        scope: example.class.name,
+        name: example.name,
+        identifier: identifier,
+        location: location,
+        file_name: file_name,
+        result: result,
+        failure_reason: failure_reason,
+        failure_expanded: failure_expanded,
+        history: history,
+      ).with_indifferent_access.compact
```
[the remark this answers is not in the document]
Comment (blaknite) [reply lib/buildkite/collector/minitest_plugin/trace.rb:47]: All good, this is an existing dependency.

**Comment by blaknite** [comment]
> Just want to mention that we probably want to include the token in that ^

Only if they're not using the default env `BUILDKITE_ANALYTICS_TOKEN`. We read that if the token is not passed.

@swebb `hook:` currently defaults to `:rspec` but maybe this should become a required argument?

**Comment by swebb** [comment]
> > Just want to mention that we probably want to include the token in that ^
> 
> Only if they're not using the default env `BUILDKITE_ANALYTICS_TOKEN`. We read that if the token is not passed.
> 
> @swebb `hook:` currently defaults to `:rspec` but maybe this should become a required argument?

I think we might be approaching bike shedding so I don't really mind. Here's my opinion but feel free to ignore 🙂

I think we should be consistent with the configuration of the token in both rspec & minitest. I don't mind if you specify the `token:` argument or use the environment variable as long as they are consistent.

I think we should specify `hook: :rspec`. That way if someone accidentally copies and pasts the rspec instructions into a minitest suite it will be really obvious what the mistake was.

**Review by JuanitoFatas — APPROVED** [review]
I’ve verified this branch works on a production Minitest app 👍🏼 

Excellent.

**Pull request merged**